File: julia_version_ENT3C/ENT3C.jl
using Pkg,  DataFrames, BenchmarkTools, JSON, Printf, Plots, ColorSchemes, SuiteSparse

include("JULIA_functions/ent3c_functions.jl")


function main()
   
    installs, resolve_env, config_file, julia_version = parse_args(ARGS)
    config = get_config(config_file)

    #args = ["--config-file=config/config.json", "--install-deps=no","--julia-version=1.11.2"]
    #installs,  resolve_env, config_file, julia_version = parse_args(args)
    required_packages = ["DataFrames", "BenchmarkTools", "JSON", "Printf", "Plots", "ColorSchemes", "SuiteSparse", "HDF5", "NaNStatistics", "Statistics", "Combinatorics", "CSV"]
    missing_packages = check_missing_packages(required_packages)

    allowed_bool = Set(["yes", "true", "no", "false"])
    check_input_option("install-deps", installs, allowed_bool)
    check_input_option("resolve-env", resolve_env, allowed_bool)

    if !isnothing(julia_version) && julia_version != ""
        allowed_versions = Set(["1.10.4", "1.11.2"])
        if !(julia_version in allowed_versions)
            println(ArgumentError("ERROR: Illegal value for --julia-version: '$julia_version'. Allowed values are: $(collect(allowed_versions))"))
            exit(1)
        end
    end

    if isnothing(installs) && isnothing(resolve_env) && !isempty(missing_packages)
        println("ERROR: Missing julia packages. Please specify one of the following:")
        println("1. Use --install-deps=yes to install the packages globally.")
        println("   Or")
        println("2. Specify --resolve-env=yes with --julia-version=1.10.4 or 1.11.2 to install and resolve environment.")
        exit(1)
    end
    
    if (installs == "yes"||installs == "true") && length(missing_packages)!=1 && (resolve_env != "yes"||resolve_env != "true")
        println("Installing missing dependencies globally.")
        println(resolve_env)
        install_packages(missing_packages)
    end
    
    
    if (resolve_env == "yes"||resolve_env == "true")
            if isnothing(julia_version) || julia_version == ""
                println(ArgumentError("--resolve-env requires julia version specification --julia-version=1.11.2 or --julia-version=1.10.4"))
                exit(1)
            end
            println("Resolving ENT3C julia enviornment.")
            Pkg.activate("project_files/$(julia_version)")
            Pkg.add(["DataFrames", "BenchmarkTools", "JSON", "Printf", "Plots", "ColorSchemes", "SuiteSparse", "HDF5", "NaNStatistics", "Statistics", "Combinatorics", "CSV"])
            Pkg.instantiate()
    end    

    
	SUB_M_SIZE_FIX = config["SUB_M_SIZE_FIX"]
	PHI_MAX = config["PHI_MAX"]
	Resolutions = config["Resolution"]
	Resolutions = split(Resolutions, ",")
	Resolutions = [parse(Float64, res) |> Int for res in Resolutions]
	phi = config["phi"]
	NormM = config["NormM"]
	DATA_PATH = config["DATA_PATH"]
	FILES = config["FILES"]
	FILES = hcat(joinpath.(DATA_PATH, FILES[1:2:end-1]), FILES[2:2:end])
	OUT_PREFIX = config["OUT_PREFIX"]
	CHRSPLIT = config["CHRSPLIT"]
	weights_name = config["WEIGHTS_NAME"]
	
	if OUT_PREFIX === nothing
	    OUT_PREFIX = @sprintf("%dkb",Resolution/1e3)
	end
	OUT_DIR = config["OUT_DIR"]
	OUT_DIR = string(OUT_DIR,"JULIA")
	if !isdir(OUT_DIR)
	    mkpath(OUT_DIR)
	end
	
	ChrNrs = string(config["ChrNr"])
	if contains(ChrNrs, ",")
	    ChrNrs = split(ChrNrs, ",")
	else
	    ChrNrs = [ChrNrs]
	end

    
    ENT3C_OUT = DataFrame(Name=Vector{String}[], ChrNr=Vector{Int}[], Resolution=Vector{Int}[],
        n=Vector{Int}[], PHI=Vector{Int}[], phi=Vector{Int}[], binNrStart=Vector{Int}[],
        binNrEnd=Vector{Int}[], START=Vector{Int}[], END=Vector{Int}[], S=Vector{Float64}[])


    for Resolution in Resolutions
        for ChrNr in ChrNrs
            FNs = []
            for f in 1:Int(size(FILES, 1))
                FNs = vcat(FNs, INFO(FILES[f, 1], FILES[f, 2]))
            end
 
            ##############################################################################
            # extract common empty bins of input matrices
            ##############################################################################
            EXCLUDE = 0
            for f in FNs
                FN = f.FN
                _, BIN_TABLE = load_cooler(FN, ChrNr, Resolution, NormM, weights_name)
                if NormM==0
                    EXCLUDE = vcat(EXCLUDE, BIN_TABLE.binNr[isnan.(BIN_TABLE.CONTACT)])
			    elseif NormM==1
				    EXCLUDE = vcat(EXCLUDE, BIN_TABLE.binNr[isnan.(BIN_TABLE.CONTACT)], BIN_TABLE.binNr[isnan.(BIN_TABLE.weights)])
			    end
            end
            EXCLUDE = unique(EXCLUDE)
            
            ##############################################################################
            #produce ENT3C_OUT data frame
            ##############################################################################
            for f in FNs
                
                FN = f.FN
                print(FN,"\n")
                M, BIN_TABLE = load_cooler(FN, ChrNr, Resolution, NormM, weights_name)

                INCLUDE = collect(1:size(M, 1))
                INCLUDE = setdiff(INCLUDE, EXCLUDE)
                M = M[INCLUDE, INCLUDE]

                BIN_TABLE = BIN_TABLE[INCLUDE, :]

                S, SUB_M_SIZE1, PHI_1, phi_1, BIN_TABLE_NEW = vN_entropy(M, SUB_M_SIZE_FIX, CHRSPLIT, PHI_MAX, phi, BIN_TABLE, FN)

                print(Resolution, " ", ChrNr, " ", FN, "\n")
                N = length(S)
                
                OUT1 = DataFrame(Name=fill(f.META, N), ChrNr=fill(ChrNr, N),
                    Resolution=fill(Resolution, N), n=fill(SUB_M_SIZE1, N),
                    PHI=fill(PHI_1, N), phi=fill(phi_1, N), binNrStart=BIN_TABLE_NEW[:, 1],
                    binNrEnd=BIN_TABLE_NEW[:, 2], START=BIN_TABLE_NEW[:, 3], END=BIN_TABLE_NEW[:, 4], S=S)

                ENT3C_OUT = vcat(ENT3C_OUT, OUT1)

                print(ENT3C_OUT,"\n")

            end
        end
    end


	#############################################################################
	# ENT3C table
	##############################################################################
	#f = [filter(row -> isnan(row.S)  , ENT3C_OUT).binNrStart[1], filter(row -> isnan(row.S)  , ENT3C_OUT).binNrEnd[1]]
	#filter(row -> row.binNrStart==f[1] && row.binNrEnd==f[2], ENT3C_OUT)
	#############################################################################
	# similarity table
	#############################################################################
	Biological_replicates = any(x -> contains(x, '_'), FILES[2, :])
	
	SAMPLES::Array=unique(ENT3C_OUT.Name)
	if length(SAMPLES)>1 
	        Similarity = get_similarity_table(ENT3C_OUT,ChrNrs,Resolutions,Biological_replicates, OUT_DIR, OUT_PREFIX)
	        CSV.write(@sprintf("%s/%s_ENT3C_similarity.csv",OUT_DIR,OUT_PREFIX), Similarity)
	end
	CSV.write(@sprintf("%s/%s_ENT3C_OUT.csv",OUT_DIR,OUT_PREFIX), ENT3C_OUT)
		
	return ENT3C_OUT, Similarity
end


ENT3C_OUT, Similarity = main()

File: matlab_version_ENT3C/ENT3C.m
function [] = ENT3C(config_file)
addpath('MATLAB_functions/')
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% load json file
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%config_file='config/config.json';
config = fileread(config_file);
config = jsondecode(config);
PHI_MAX=config.PHI_MAX;
CHRSPLIT=config.CHRSPLIT;
SUB_M_SIZE_FIX=config.SUB_M_SIZE_FIX;
Resolutions=str2num(config.Resolution);
phi=config.phi;
NormM=config.NormM;
weights_name=config.WEIGHTS_NAME;
DATA_PATH=config.DATA_PATH;
FILES = config.FILES;
FILES = reshape(FILES,2,size(FILES,1)/2);
FILES = [cellfun(@(file) fullfile(DATA_PATH, file), FILES(1,:), 'UniformOutput', false); FILES(2,:)];
if any(contains(FILES(2,:),'_'))% only important in case of computing mean(Q_BRs) mean(Q_nonBRs)
    Biologicap_replicates=true;
else
    Biologicap_replicates=false;
end
OUT_DIR=config.OUT_DIR;
OUT_DIR=[OUT_DIR,'/MATLAB'];
OUT_PREFIX=config.OUT_PREFIX;
if isempty(OUT_PREFIX)
    OUT_PREFIX=sprintf('%dkb',Resolution/1e3);
end
if ~exist(OUT_DIR, 'dir')
    mkdir(OUT_DIR);
end

ChrNrs=config.ChrNr;
ChrNrs=strsplit(ChrNrs, ',');

ENT3C_OUT_FN=sprintf('%s/%s_ENT3C_OUT.csv',OUT_DIR,OUT_PREFIX);

if ~exist(ENT3C_OUT_FN,'file')
    sprintf("Generating new file: %s",ENT3C_OUT_FN)
    ENT3C_OUT=table('Size',[0 11],...
    'VariableTypes', {'cell','cell','int64','int64','int64','int64','int64','int64','int64','int64','double'},...
    'VariableNames', {'Name','ChrNr','Resolution','n','PHI','phi','binNrStart','binNrEND','START','END','S'});
    for Resolution=Resolutions
        for ChrNr=ChrNrs

            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            % load files
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            INFO = struct('FN', '', 'META', '');
            FNs = [];
            for f = 1:size(FILES,2)
                FNs = [FNs, INFO];
                FNs(end).FN = FILES{1,f};
                FNs(end).META = FILES{2,f};
            end
            CM=jet(size(FNs,2));
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            % extract common empty bins of input matrices
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            EXCLUDE=[];
            for f = 1:numel(FNs)
                FN=FNs(f).FN;
                [M,BIN_TABLE]=load_cooler(FN,ChrNr,Resolution,NormM,weights_name);
                if NormM==0
                    EXCLUDE=[EXCLUDE;BIN_TABLE.binNr(isnan(BIN_TABLE.CONTACT))];
                elseif NormM==1
                    EXCLUDE=[EXCLUDE;BIN_TABLE.binNr(isnan(BIN_TABLE.CONTACT));BIN_TABLE.binNr(isnan(BIN_TABLE.weights))];
                end
            end
            EXCLUDE=unique(EXCLUDE);
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            % fill ENT3C table
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            % figure(1);tiledlayout(2,2,'tilespacing','compact');
            for f = 1:numel(FNs)
                FN=FNs(f).FN;
                [M,BIN_TABLE]=load_cooler(FN,ChrNr,Resolution,NormM,weights_name);
                INCLUDE = 1:size(M,1);
                INCLUDE = setdiff(INCLUDE,EXCLUDE);

                M = M(INCLUDE,INCLUDE);

                %{
            PERC_0 = 100*sum(isnan(M(:)))/numel(M(:))
            nexttile
            imagesc(M);imagesc(log(M));axis square;set(gca,'YDir','normal');title(sprintf('Chr%s %s NormM%d %d%%',...
                    ChrNr{1},FNs(f).META,NormM,round(100*sum(isnan(M(:)))/numel(M(:)))))
                %}
                BIN_TABLE = BIN_TABLE(INCLUDE,:);

                % writetable(BIN_TABLE,sprintf('%s_chr%d_BINMATRIXMATLAB.csv',FN,ChrNr),'Delimiter','tab')

                [S, SUB_M_SIZE1, PHI_1, phi_1, BIN_TABLE_NEW] = vN_entropy(M,SUB_M_SIZE_FIX,CHRSPLIT,PHI_MAX,phi,BIN_TABLE);
                N = length(S);

                OUT1 = table(repmat({FNs(f).META},N,1),...
                    repmat({sprintf('%s',ChrNr{1})},N,1),repmat(Resolution,N,1),...
                    repmat(SUB_M_SIZE1,N,1),repmat(PHI_1,N,1),repmat(phi_1,N,1),...
                    BIN_TABLE_NEW(:,1),BIN_TABLE_NEW(:,2),...
                    BIN_TABLE_NEW(:,3),BIN_TABLE_NEW(:,4),S,...
                    'VariableNames',{'Name','ChrNr','Resolution','n','PHI','phi',...
                    'binNrStart','binNrEND','START','END','S'});

                ENT3C_OUT=[ENT3C_OUT;OUT1];
              
            end
        end
    end

    writetable(ENT3C_OUT,ENT3C_OUT_FN,'Delimiter','tab')

else

    opts = detectImportOptions(ENT3C_OUT_FN, 'TextType', 'string');
    opts = setvartype(opts, 'ChrNr', 'string'); % or 'char'
    ENT3C_OUT=readtable(sprintf('%s',ENT3C_OUT_FN),opts);
end


%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fill similarity table
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
SAMPLES=unique(ENT3C_OUT.Name);

Similarity = [];
if numel(SAMPLES)>1
    comparisons = get_pairwise_combs(SAMPLES);
    for Resolution=Resolutions
        figure(Resolution);hold on;tiledlayout('flow');hold on%'flow');hold on
        for ChrNr=ChrNrs

            figure(Resolution);nexttile

            plotted = "tempstring";
            c=1;
            for f=1:size(comparisons,1)

                S1 = ENT3C_OUT(strcmp(ENT3C_OUT.Name,comparisons{f,1})&strcmp(ENT3C_OUT.ChrNr,ChrNr)&ENT3C_OUT.Resolution==Resolution,:);
                S2 = ENT3C_OUT(strcmp(ENT3C_OUT.Name,comparisons{f,2})&strcmp(ENT3C_OUT.ChrNr,ChrNr)&ENT3C_OUT.Resolution==Resolution,:);
                non_nan_idx = ~isnan(S1.S)&~isnan(S2.S);
                Q = corrcoef(S1.S(non_nan_idx),S2.S(non_nan_idx));Q=Q(1,2);

                Similarity=[Similarity;...
                    table(Resolution,ChrNr,comparisons{f,1},comparisons{f,2},Q,...
                    'VariableNames',{'Resolution','ChrNr','Sample1','Sample2','Q'})];

                if ~ismember(comparisons{f,1}, plotted)
                    figure(Resolution);plot(S1.S,'color',CM(c,:),'linewidth',1.3);hold on;axis tight
                    plotted = [plotted;comparisons{f,1}];c=c+1;
                end
                if ~ismember(comparisons{f,2}, plotted)
                    figure(Resolution);plot(S2.S,'color',CM(c,:),'linewidth',1.3);hold on;axis tight
                    plotted = [plotted;comparisons{f,2}];c=c+1;
                end
            end
            if Biologicap_replicates
                title(sprintf('Chr%s %dkb\n$\\overline{Q}_{BR}=%4.2f$ $\\overline{Q}_{nonBR}=%4.2f$',...
                    ChrNr{1},Resolution/1e3,...
                    mean(Similarity.Q(strcmp(Similarity.ChrNr,ChrNr{1})&Similarity.Resolution==Resolution&...
                    strcmp(extractBefore(Similarity.Sample1,'_'),extractBefore(Similarity.Sample2,'_')))),...
                    mean(Similarity.Q(strcmp(Similarity.ChrNr,ChrNr{1})&Similarity.Resolution==Resolution&...
                    ~strcmp(extractBefore(Similarity.Sample1,'_'),extractBefore(Similarity.Sample2,'_'))))),...
                    'interpreter','latex','fontsize',25)
            else
                title(sprintf('Chr%s %dkb',ChrNr{1},Resolution/1e3),'interpreter','latex','fontsize',25)
            end

            last=ChrNrs(end);
            if strcmp(ChrNr{1},last{1})
                plotted(1)=[];plotted=strrep(plotted,'_',' ');
                legend(plotted,'location','northeastoutside')
            end
            set(gca,'FontSize',15)
        end

        figure(Resolution);
        set(gcf,'Position',[52 215 1717 813])
        saveas(gcf,sprintf('%s/%s_%d_ENT3C_signals.svg',OUT_DIR,OUT_PREFIX,Resolution))

    end
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    % save output
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    writetable(Similarity,sprintf('%s/%s_ENT3C_similarity.csv',OUT_DIR,OUT_PREFIX),'Delimiter','tab')
end

